File: 3.Longest_Substring/substringIntArr.java
import java.util.Arrays;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int maxLength = 0;
        int[] charIndex = new int[128];
        Arrays.fill(charIndex, -1);
        int left = 0;
        
        for (int right = 0; right < n; right++) {
            if (charIndex[s.charAt(right)] >= left) {
                left = charIndex[s.charAt(right)] + 1;
            }
            charIndex[s.charAt(right)] = right;
            maxLength = Math.max(maxLength, right - left + 1);
        }
        
        return maxLength;
    }
}

public class substringIntArr{
    public static void main(String[] args){
        Solution sol = new Solution();
        String s = "abcabcbb";
        System.out.println(sol.lengthOfLongestSubstring(s));
    }
}
// Time complexity: O(n)
// Space complexity: O(128) = O(1)
// n is the length of s

File: 3.Longest_Substring/substringIntArr.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;
class Solution{
    public:
        int lengthOfLongestSubstring(string s){
            int n = s.size();
            int ans = 0;
            vector<int> index(128, -1);
            int left = 0;
            for(int right = 0; right < n; right++){
                if(left <= index[s[right]]){
                    left = index[s[right]] + 1;
                }
                index[s[right]] = right;
                ans = max(ans, right - left + 1);
            }
            return ans;
        }
};

int main(){
    string s = "abcabcbb";
    Solution sol;
    cout << sol.lengthOfLongestSubstring(s) << endl;
    return 0;
}